### DISCUSSION TITLE
Interpretations of Dot plot output

### DISCUSSION THREAD

**Discussion by u7068758** [cogent3/cogent3#1057] [Q&A]
Can anyone guide me about the result interpretations of Dot plot? How can I best describe the results of Dot plot output in the thesis?
As per my understanding, the straight line indicates the matching amino acids and short parallel lines indicate repeats. Thus, we can say both the sequences are similar having tandem repeats at the end. At the top of output, it is written Matched>=13/20 what does this mean?
Also, Gap <=0 but according to alignment (clustal omega) there is one repeat missing in sequence on y-axis
![image](https://user-images.githubusercontent.com/107029444/172769351-8eebc316-40fc-495e-8edc-76603ee590b3.png)
![image](https://user-images.githubusercontent.com/107029444/172769527-7b43a620-76c6-40e2-8f7f-7bf13415e2f9.png)

**Comment by GavinHuttley** [comment]
There's a cluster of repeat sequences, indicated by the dense parallel lines. If you hover your mouse cursor over the plot, the sequence coordinates will be shown. So you could slice the individual sequences to get the corresponding segments and examine them.

To understand the "title" of the plot (i.e. what window, threshold, gap mean), look at help on the dotplot method. That is, in the notebook cell
```python
help(seqs.dotplot)
# or for prettier display inside a notebook you can do
seqs.dotplot?
```
As a side note, the default values are for DNA. I think those are too stringent for a comparison of protein sequences. I recommend reducing window size, but keep the relative magnitude of threshold the same.

Better yet, I strongly encourage you to do the same plot but with the corresponding DNA sequences.

On a side note, one of the best way to understand these sorts of techniques is to construct a small synthetic data set that has a property of interest. By using such small synthetic cases and modifying the settings you will be able to interpret real data in a more informed manner.

**Reply by u7068758** [reply]
As per your kind suggestion, I took a slice (62 amino-acids) of the region containing repeats and performed the plot (graph1) and then converted the amino acid sequence into a nucleotide sequence (186bp) and again performed a plot (graph#2). Please see the graphs and suggest which one is good to use. Unlike the protein plot, the plot of nucleotide sequences is showing double lines at the end. Can you guide what these additional small lines indicate? I also performed a reverse complement plot shown in graph #3
![Protein plot](https://user-images.githubusercontent.com/107029444/172791879-ad285326-c071-4faf-824f-352b1cc612dd.png)
![DNAplot](https://user-images.githubusercontent.com/107029444/172792492-975a469d-2a49-42c7-8a50-40303e7cf153.png)
![RCplot](https://user-images.githubusercontent.com/107029444/172793394-1bfee942-51bb-4278-8115-a9d86f1c58d6.png)

**Comment by GavinHuttley** [comment]
What I mean by "...construct a small synthetic data set..." was literally make a DNA sequence up. Even though there's only one sequence in this collection, a dotplot is possible since the sequence will be just compared to itself.

In the following example I have constructed a synthetic sequence that has a specific property which only manifest if the window size is exactly 4. I've set the threshold to also 4. Code is below.

```python
from cogent3 import make_unaligned_seqs

seqs = make_unaligned_seqs({"seq1": 'CACACCACTGCAGTCGGATAGACC'}, moltype="dna")
seqs.dotplot(window=4, threshold=4, rc=True)
```
I'll let you run that code yourself.

If you take a look at the sequence object itself (not the collection), you will get colour information that can assist in seeing patterns. That is, in a Jupyter notebook, you can access `.seqs` attribute of a `SequenceCollection`. In this case, there's only one sequence, we get it as
```python
seqs.seqs[0]
```
<img width="633" alt="dotplot" src="https://user-images.githubusercontent.com/3102996/173272087-46f354eb-1609-4bf2-8002-99967bba10c6.png">

Can you figure out the "specific property" I added to the sequence?

**Reply by u7068758** [reply]
Thank you Gavin,
The dot plot is below:
![newplot](https://user-images.githubusercontent.com/107029444/173412203-d6c950e8-37dc-450e-8816-3a789e7520fb.png)

It indicates the presence of a motif. Nucleotides at position 5 and 6 (CC) are complementary to those at positions 14 and 15 (GG), that may result in loop formation, interfering the replication and resulting in mutation.

**Reply by GavinHuttley** [reply]
First, this is a synthetic sequence so forget the biological explanation. 

Second, a match (blue line, or red on reverse complement) is defined as `window==threshold==4`. Thus, **all 4 bases must be identical** in the compared sequence windows. So you are right to identify the complementarity (`CC`, `GG`), but 2 nucleotides is not considered a match.

Look at what positions the reverse-complement diagonal starts and ends, then look at those positions in the sequence.

**Reply by u7068758** [reply]
The reverse complement diagonal starts at (6,14) and ends at (15,5) but nucleotides at these positions are neither identical nor complementary in a sequence. 
Nucleotides at position#6-9 (ACTG) are complement to those in position 10-13 (CAGT) in reverse direction. In other words, the nucleotides at positions (6,13), (7,12), (8,11), and (9,10) are complementary to each others.
Please clarify whether the red line indicates the identical nucleotides or complementary ones?

**Reply by GavinHuttley** [reply]
look at the legend on the plot

**Reply by GavinHuttley** [reply]
Also, you might find it helpful to look at the reverse complement sequence. You can create that as
```python
seq = seqs.seqs[0]
rc = seq.rc()
rc
```

**Reply by u7068758** [reply]
Legend says Matched >= 4/4  Gap<=0
Does it mean that the reverse complement diagonal (starting at 6,14 and ending at 15,5) contains at least four matching consecutive nucleotides? So, all the 9 nucleotides may not be matching within the diagonal.

I've looked at the reverse complement sequence before. It is showing the same behavior as I have described above in my previous reply.

**Reply by GavinHuttley** [reply]
The best way to really understand this algorithm is to do it manually yourself. Let's just focus on comparing the plus strand with the minus strand (reverse complement).

Four our example, the algorithm will takes all possible subsequences of length 4 from the plus strand sequence and look for all exact matches on the minus-strand ( the reverse complemented) sequence.

Before "starting" our algorithm, let's be clear on how a match is translated into a "dot" on the plot. The x-coordinate is the start position of the plus-strand sequence window. For the y-coordinate, subtract the minus-strand start of the matching window from the length of the sequence.

Now start the algorithm by selecting the first four bases (positions 0, 1, 2, 3) of the plus strand sequence. Look at the reverse complement, is there a match anywhere? If yes, determine the coordinates and put a mark on your plot. Select the next four (positions 1, 2, 3, 4). Look at the reverse complement, is there a match anywhere? etc..

Continue until your plus strand window is the last four positions of the plus strand sequence.

**Reply by u7068758** [reply]
Thank you Gavin, for guiding the procedure.
I'm not sure whether to subtract from 24 (there are 24 bp) or 23 (positions are 0-23)

In case I subtract from 24, the positions 6-10 (ACTGC) on + strand are identical to positions 10-14 (ACTGC)on reverse complement strand. After subtraction from sequence length (24), the resultant coordinates are (6,14), (7,13),(8,12),(9,11),and (10,10). The plot gives a line same as that of red line in cogent3 graph but shorter in length.

If we look at first coordinate (6,14) what does this indicate? 6 is a position on plus strand. I have confusion about 14. The nucleotide at position 14 in both + and - strands are not matching to position 6. Then what does this 14 indicate in this case?

My best guess is_in case of (6,14)_ we have to subtract 14 from 24, the result i.e. 10 indicates the position on reverse complement strand. Thus, "A" at position 6 on plus strand is matching to "A" on position 10 on reverse complement strand. 

In case we subtract from 23 (as there are 0-23 positions), then the resulting values are  (6,13), (7,12), (8,11), and (9,10) and residues in these pairs are complementary.

**Reply by GavinHuttley** [reply]
To draw a line on a Cartesian plane (a graph) you need 2 coordinates. In a dotplot, these coordinates correspond to positions on the sequences being compared. The resulting line is a regions that matches between the two sequences, where the a match is defined by the window/threshold values.

There is a bug in the displayed reverse complement (it extends 1 base too far).

Here's a demonstration that the coordinates should be coord1 = (6, 24-10), coord2 = (14, 24-18) :
```python
from cogent3 import make_seq

s = make_seq('CACACCACTGCAGTCGGATAGACC', moltype="dna")
r = s.rc()
s[6:14] == r[10:18]
```

I'm sorry, but I can't continue this discussion as I have run out of time. Good luck!

**Reply by u7068758** [reply]
Thank you Gavin, I learned a lot from this detailed discussion.
